gitprovider.go

```diff
@@ -0,0 +1,115 @@
+
+	// ErrUnauthorized describes that that the request was unauthorized by the Git provider server
+	ErrUnauthorized = errors.New("unauthorized request for this resource")
+	// ErrRateLimited is returned if a request hit the rate limit of the Git provider server
+	ErrRateLimited = errors.New("hit rate limit")
+
+	// ErrNotTopLevelOrganization describes the case where it's mandatory to specify a top-level organization
+	// (e.g. for accessing teams), but a sub-organization was passed as the OrganizationRef
+	ErrNotTopLevelOrganization = errors.New("the specified organization wasn't a top-level one as expected")
+
+	// ErrAlreadyExists is returned by .Create() requests if the given resource already exists.
+	// Use .Reconcile() instead if you want to idempotently create the resource
+	ErrAlreadyExists = errors.New("resource already exists, cannot create object. Use Reconcile() to idempotently create.")
+	// ErrNotFound is returned by .Get() and .Update() calls if the given resource doesn't exist
+	ErrNotFound = errors.New("the requested resource was not found")
+
+	// ErrFieldRequired specifies the case where a required field isn't populated at use time
+	ErrFieldRequired = errors.New("field is required")
+	// ErrFieldInvalid specifies the case where a field isn't populated in a valid manner
+	ErrFieldInvalid = errors.New("field is invalid")
+	// ErrFieldEnumInvalid specifies the case where the given value isn't part of the known values in the enum
+	ErrFieldEnumInvalid = errors.New("field value isn't known to this enum")
+
+	// ErrURLUnsupportedScheme is returned if an URL without the HTTPS scheme is parsed
+	ErrURLUnsupportedScheme = errors.New("unsupported URL scheme, only HTTPS supported")
+	// ErrURLUnsupportedParts is returned if an URL with fragment, query values and/or user information is parsed
+	ErrURLUnsupportedParts = errors.New("organization or repository URLs cannot have fragments, query values nor user information")
+	// ErrURLInvalid is returned if an URL is invalid when parsing
+	ErrURLInvalid = errors.New("invalid organization or repository URL")
+	// ErrURLMissingRepoName is returned if there is no repository name in the URL
+	ErrURLMissingRepoName = errors.New("missing repository name")
+)
+
+// ProviderID is a typed string for a given Git provider
+// The provider constants are defined in their respective packages
+type ProviderID string
+
+// Creator is an interface which all objects that can be created
+// (using the Client) should implement
+type Creator interface {
+	// ValidateCreate will be run in every .Create() client call, before defaulting
+	// Set (non-nil) and required fields should be validated
+	ValidateCreate() error
+	// Default will be run after validation, setting optional pointer fields to their
+	// default values before doing the POST request
+	Default()
+}
+
+// Creator is an interface which all objects that can be updated
+// (using the Client) should implement
+type Updator interface {
```

Comment:
[Updator vs updater](https://english.stackexchange.com/questions/16161/one-who-creates-is-a-creator-what-is-one-who-updates), updater seems to be the more correct term of the two. I'd also argue that this should be `Updatable`, since it describes the capability of receiving an update.